File: arrays/LC88.py
class OldSolution(object):
    def merge(self, nums1, m, nums2, n):
        """
        :type nums1: List[int]
        :type m: int
        :type nums2: List[int]
        :type n: int
        :rtype: None Do not return anything, modify nums1 in-place instead.
        """
        temp = 0
        num = 0
        for x in range(0, len(nums1)):
            if n == 0:
                break
            if nums1[x] > nums2[-n]:
                num = nums1[x]
                nums1[x] = nums2[-n]
                n = n - 1
                # Need to shift all remaining vals one idx to the right
                for y in range(0, m):
                    temp = nums1[x+y+1]
                    nums1[x+y+1] = num
                    num = temp
            elif m == 0:
                nums1[x] = nums2[-n]
                n = n - 1
            else:
                m = m - 1


class AlmostSolution(object):
    def merge(self, nums1, m, nums2, n):
        """
        :type nums1: List[int]
        :type m: int
        :type nums2: List[int]
        :type n: int
        :rtype: None Do not return anything, modify nums1 in-place instead.
        """
        if n == 0:
            return

        p1 = 0
        p2 = 0
        idx = 0

        nums1_part = list(nums1[:m])

        while (p1 < m or p2 < n) and idx < m+n:
            if p1 < m and p2 < n:
                if nums2[p2] < nums1_part[p1]:
                    nums1[idx] = nums2[p2]
                    p2 += 1
                else:
                    nums1[idx] = nums1_part[p1]
                    p1 += 1

            elif p1 < m:
                nums1[idx] = nums1_part[p1]
                p1 += 1
            else:
                nums1[idx] = nums2[p2]
                p2 += 1

            idx += 1

        
class OptimalSolution(object):
    def merge(self, nums1, m, nums2, n):
        """
        :type nums1: List[int]
        :type m: int
        :type nums2: List[int]
        :type n: int
        :rtype: None Do not return anything, modify nums1 in-place instead.
        """
        if n == 0:
            return

        p1 = m-1
        p2 = n-1
        idx = m+n-1

        while (p1 > -1 or p2 > -1) and idx > -1:
            # print(nums1)
            if p1 > -1 and p2 > -1:
                if nums2[p2] > nums1[p1]:
                    nums1[idx] = nums2[p2]
                    p2 -= 1
                else:
                    nums1[idx] = nums1[p1]
                    p1 -= 1

            elif p1 > -1:
                nums1[idx] = nums1[p1]
                p1 -= 1
            else:
                nums1[idx] = nums2[p2]
                p2 -= 1

            idx -= 1

        
'''
Revisited this problem and initially came up with the O(m+n) RT, O(m) space solution
"AlmostSolution". I then realized doing almost the exact same thing, but starting from
the ends of the arrays would mean using O(1) space, which lead me to "OptimalSolution", 
which has a runtime of O(m+n), and a space complexity of O(1)

Here's a similar discussion board solution

class Solution:
    def merge(self, nums1, m, nums2, n):
        midx = m - 1
        nidx = n - 1 
        right = m + n - 1

        while nidx >= 0:
            if midx >= 0 and nums1[midx] > nums2[nidx]:
                nums1[right] = nums1[midx]
                midx -= 1
            else:
                nums1[right] = nums2[nidx]
                nidx -= 1

            right -= 1
'''
            
        
        

File: arrays/LC88.cpp
#include <vector>

class Solution {
public:
    void merge(std::vector<int>& nums1, int m, std::vector<int>& nums2, int n) {
        if(n == 0) {
            return;
        }
        if(m == 0) {
            nums1 = nums2;
            return;
        }
        int num1_idx = m - 1;
        int num2_idx = n - 1;
        

        for(int repl = m + n - 1; repl > -1; repl--) {
            if(num2_idx < 0) {
                nums1[repl] = nums1[num1_idx];
                num1_idx -= 1;
            } else if(num1_idx < 0) {
                nums1[repl] = nums2[num2_idx];
                num2_idx -= 1;
            } else if(nums2[num2_idx] > nums1[num1_idx]) {
                nums1[repl] = nums2[num2_idx];
                num2_idx -= 1;
            } else {
                nums1[repl] = nums1[num1_idx];
                num1_idx -= 1;
            }
        }

    }
};

/* 
RT: O(n + m)
Space: O(1)

This an optimal solution! However, while loops are more effective:

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int midx = m - 1;
        int nidx = n - 1;
        int right = m + n - 1;

        while (nidx >= 0) {
            if (midx >= 0 && nums1[midx] > nums2[nidx]) {
                nums1[right] = nums1[midx];
                midx--;
            } else {
                nums1[right] = nums2[nidx];
                nidx--;
            }
            right--;
        }        
    }
};
*/